### ISSUE TITLE
an assert function forcing compilation to quit

### ISSUE THREAD

**Issue by ivoknutsel** [openscad/openscad#381]
With large scripts and many parameters certain combinations of parameters may not generate a usable result. Currently there is no way to quit compilation and return a string describing the problem.

I propose an assert function with a syntax like : assert(<test if true>, <return string>);

If the <test if true> value is false, the function will abort generation and return the <return string> value. The return string value can be constructed like the values for echo.

**Comment by 5solids** [comment]
+1

I have the same need. In the beginning of my script there are a number of configuration variables the user can modify. However some combinations can cause problems, so I would like to check them and stop processing right there.

What I am currently doing is to write an error message to the console (using the echo() function and formatting the text with HTML to be bold and red), but the object still gets built.

I also tried to explicitly cause an error (e.g. division by zero, sqrt(-1), include/use <_none_existing_file> and similar), but none of them cause the processing to stop. The first two didn't even raise a warning.

@GilesBathgate Can you please elaborate a bit on this?

**Comment by TakeItAndRun** [comment]
How about:

// define r here
r=15;
// r=32;

function no_error()=(r>=30)?false:true;

if(no_error()) myPart();

module myPart(){
...
}

2015-01-12 20:58 GMT+01:00 5solids <notifications@github.com>:

> +1
>
> I have the same need. In the beginning of my script there are a number of
> configuration variables the user can modify. However some combinations can
> cause problems, so I would like to check them and stop processing right
> there.
>
> What I am currently doing is to write an error message to the console
> (using the echo() function and formatting the text with HTML to be bold and
> red), but the object still gets built.
>
> I also tried to explicitly cause an error (e.g. division by zero,
> sqrt(-1), include/use <_none_existing_file> and similar), but none of them
> cause the processing to stop. The first two didn't even raise a warning.
>
> @GilesBathgate <https://github.com/GilesBathgate> Can you please
> elaborate a bit on this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openscad/openscad/issues/381#issuecomment-69634737>.
>



-- 
stempeldergeschichte@googlemail.com <karsten@rohrbach.de>

P.S. Falls meine E-Mail kürzer ausfällt als Dir angenehm ist:
Ich probiere gerade aus kurze Antworten statt gar keine Antworten zu
schreiben.
Wenn Du gerne mehr lesen möchtest, dann lass es mich bitte wissen.

P.S. In case my e-mail is shorter than you enjoy:
I am currently trying short replies instead of no replies at all.
Please let me know, if you like to read more.

Enjoy!

**Comment by MichaelAtOz** [comment]
or

module check() {
if (your_tests)
  children();
}

check() your_part();

But I +1 the OP too, I tried all the crafty ways to kill it too, it is robust...

**Comment by 5solids** [comment]
Thanks for the two examples. I like that they are a way to prevent the creation of the objects in case a check went wrong.

However in my script I have to check multiple things and also want to provide an error message for each individual check. So yes, the example code will work there too, but the overall code structure becomes a bit messy:

foo = 23;
bar = 42;

valid_foo = (foo == 42) ? true : false;
valid_bar = (bar == 42) ? true : false;
if (!valid_foo) {
	echo("ERROR: Invalid value of foo variable.");
}
if (!valid_bar) {
	echo("ERROR: Invalid value of bar variable.");
}
valid_config = (valid_foo && valid_bar) ? true : false; 

module main() {
	// do something with foo & bar
}
if (valid_config) {
	main();
}

Note that I replaced the check functions with variables to increase readability (eliminating the brackets). Any downside to this?

So overall I think this is the work-around I will be using for now. However an assert(expr) or a exit(rc) would still be highly appreciated.

**Comment by greysAcademicCode** [comment]
+1

**Comment by Skeen** [comment]
We (at IOMOSS) are using the following as an assert function.
It echos the provided error-message, and then terminates the compilation.
Compilation is terminated by invoking infinite recursion.

```
// Called when an assertion fails (invokes infinite recursion).
function assertion_failed() = (assertion_failed());

// Assert that `bool` is true.
//
// If `bool` is false, emit the error `msg`,
//  and terminate compilation (via infinite recursion).
module assert(bool, msg = "")
{
    if(bool == false)
    {
        // Start spam recursion
        echo("Assertion Failed: ", msg);
        echo("", assertion_failed());
    }
}
```

**Comment by doug-moen** [comment]
I would like an `assert(condition)` operation that can be used either in modules or in functions. It prints a message and aborts the script if the condition is false.

It's okay if there is an optional second argument for specifying a message, but I don't want to be forced to use this. I want the implementation to provide a useful message by default, if none is given, similar to the assertions found in other programming languages, and especially the assertions found in unit test packages, which often give excellent and informative failure messages.

Statement syntax: `assert(condition);`

The expression syntax is similar to a `let` expression: `assert(condition) value_if_true`.

The type test functions of #1584 will be useful when combined with assertions. For example, here is a wrapper around the `cube` module which reports errors if bad arguments are passed:
```
module cuboid(size,center=false)
{
   assert(is_num(size) || is_vec3(size));
   assert(is_bool(center));
   cube(size, center);
}
function is_vec3(V) = is_list(V) && len(V)==3 && is_num(V[0]) && is_num(V[1]) && is_num(V[2]);
```
Here's a wrapper around the `len` function which checks its argument:
```
function length(X) =
   assert(is_list(X) || is_string(X))
   len(X);
```

In issue  #1574, we want OpenSCAD to abort with a message if a built-in module or function is called with bad arguments. This issue is related, because here we want the ability to write our own modules and functions that abort with a message if it is called with bad arguments.

**Comment by GilesBathgate** [comment]
Is this an improvement on issue https://github.com/openscad/openscad/issues/381

**Comment by doug-moen** [comment]
Giles, what? This is issue #381.

**Comment by doug-moen** [comment]
Here are some suggestions on how `assert` could generate good error messages for failed assertions, based on my experience with programming languages and unit test frameworks.

* `assert(condition)` prints the source code of the condition if the assertion fails.
* If the top level operator of the condition is `&&`, then only print the source code for the subcondition which is false, instead of printing the entire condition.
* Print the values of any local variables or named arguments that are referenced in the condition.
* Print the file name and line number of the failed assertion.
* Instead of just printing the file name and line number, print a stack trace.

**Comment by GilesBathgate** [comment]
@doug-moen Yes I realised as soon as I posted, you are too quick off the ball ;)

**Comment by t-paul** [comment]
See pull request #1587 for a first implementation of `assert()`. We don't have the data yet to give line numbers or variable values at this point. That's certainly needed but it's a separate bigger issue.

We don't need any `&&` trickery as the default syntax allows for multiple expressions that can be evaluated separately.

**Comment by MichaelAtOz** [comment]
Is `assert()` a function only, can't be a statement?

In your example `v = assert(a < 20, test2 = a < 10, test3 = b == 20, test4 = b != 20 && "Test! <html>&");`

What purpose does `v` serve, just a place holder? What value does it get?

Is the `&&` a plain old AND, and as non-empty strings are TRUE, is that an example of how to specify the text message (ie as part of the logical expression) for the test4 condition?

**Comment by t-paul** [comment]
Don't worry, that's just the expression example :-). It can be a statement also, and (possibly) the same for `echo()` too. For the `echo()` statement change, we have to consider the backward compatibility though. I hope it's ok, as pretty much all the new stuff would have no effect before.

`assert(a < 10) cube(a);`

...and oops, the function case misses one important bit, the "return value".

`v = assert(a < 10, b < 10) a + b;`

**Comment by doug-moen** [comment]
Given a choice between encouraging people to write
```
assert(condition && "message")
```
and supporting an explicit message argument like this:
```
assert(condition, msg="message")
```
I would prefer the latter. It's a cleaner interface, and it would provide more flexibility for formatting the message string.

Also, in your example, I don't understand the purpose of the labeled arguments 'test2=...' etc.

**Comment by MichaelAtOz** [comment]
> I would prefer the latter. 

+1 for simple cases & in terms of the language style, but that should not prevent the former. I can see the `test2=...` (along with && "some msg") as being useful for a complex set of tests. Much like `echo(var=var)` is used as a shortcut to output `var=the_value` it provides the variable name (or in the assert context the test name), without having to do a bunch of str()'s and concatenations.

Think about the longest recursion function you can imagine, and then needing to stick in multiple echo()'s and/or assert()'s to debug it. Having labels would be rather handy I'd imagine.

**Comment by t-paul** [comment]
We can also have `assert(condition = <..>, message = "message on failure");"` but then we can't have multiple separate checks where only the failure part is shown (I'm voting against printing parts of a `&&` expression). This also will not prevent the usage of `cond && "text"`. I guess the multiple checks case could be handled by simply using multiple `assert()` statements/expressions.

So with that API, there's no new scope created that holds the call arguments, hence the message will also not be available inside the following code.

**Comment by doug-moen** [comment]
@MichaelAtOz said "Think about the longest recursion function you can imagine, and then needing to stick in multiple echo()'s and/or assert()'s to debug it. Having labels would be rather handy I'd imagine."

Here's the problem. I'd like an implementation of assert that doesn't require me to include messages and labels. I don't want to spend time iteratively cluttering up my assertions with labels and messages until there is enough information printed so that I can distinguish one assertion failure from another. You're telling me there are two different kludgy ways to label asserts, and that I will need to use them.

The problem is that the assertion failure message contains no context, other than the condition that evaluated to false. There must be some additional context that we can print.

Here are some ideas.

In parser.y, when we construct the parse tree node for an assertion, store the file name and line number of the assertion. When the assertion fails, print the file name and line number. Based on Torsten's code, we'd add file,line arguments to ExpressionFunctionCall::create().

Print a stack trace when an assertion fails. In principle, we should be able to print filename, linenumber, function name, arguments, for each function call in the call stack.
* The `Context` structure already contains a parent pointer, so we could walk the Context chain to produce the stack trace. We'll need to add another data member to Context that is non-null for function call contexts, which contains stack trace information for that function call. This new member will be an `EvalContext*`.
* An `EvalContext` holds the evaluation context for a function or module call. So add stack trace information to this structure (file name, line number, function or module name, or a pointer to an object that contains this information). `ExpressionFunctionCall::evaluate` creates the EvalContext for a function call, and the ExpressionFunctionCall object contains this information.
* `Function::evaluate` constructs a context `c`, then calls c.setVariables with the EvalContext as the second argument. Modify Context::setVariables to store a pointer to the EvalContext in the Context structure; this marks the Context as a function call context, and provides access to the information needed to print the stack trace when we walk the Context chain.

In order to print the values of variables referenced by the false condition, we'd need to add a virtual printVariables() member function to Expression, which is passed a Context. The default implementation, Expression::printVariables, just calls this function recursively on its children. ExpressionLookup::printVariables overrides the default implementation, and prints name=value. That's pretty crude, because the same variable might get printed multiple times, but we could clean that up by remembering which variables have been printed already.

I've never actually modified the OpenSCAD evaluator before, so I may have misunderstood how the data structures work, and there may be a better way.

**Comment by GilesBathgate** [comment]
@t-paul When the assertion happens, does OpenSCAD print a stack trace of the functions and modules in the script?

**Comment by doug-moen** [comment]
@t-paul said "I'm voting against printing parts of a && expression."

Yes and no.

In every other programming language, you write `assert(condition)`, and writing `assert(c1 && c2)` is normal, and doesn't cause a problem. We should optimize the design of `assert` for this use case, since that's how people will tend to use it.

My suggestion of only printing the false subconditions of an && condition is an optimization that increases the amount of information printed by an assertion failure in the normal case, where people write a single condition: now you know which subcondition was false. (And yes, I did implement this feature for a unit test framework I once wrote. The design was based on feedback from about 6 other users.)

But it's only an improvement if we also print the context of the assertion failure, as I've suggested (stack trace, variables, file name/line number). If there is no context, then printing false subconditions of an && condition might be worse, since the argument to assert is the only information you are given for determining which assertion failed.

**Comment by t-paul** [comment]
As mentioned above, right now we lack the framework to give additional information, and that's also a bigger topic. I'd prefer to give the limited but still quite useful feature now instead of making it a huge project that will never happen due to time constraints.
Improving on error reporting is a general thing anyway and not limited to just `assert()`.

**Comment by GilesBathgate** [comment]
@t-paul That's fair enough. Hopefully the design it future compatible and allows this to be implemented retrospectively. Although I did note that the implementation uses C++ exceptions to facilitate ease of implementation. (I.e. getting out of the c++ evaluation stack easily) does this pose a problem in generating call stacks, and/or user-space exception catching?

**Comment by doug-moen** [comment]
@t-paul Any feedback on my proposed implementation of a stack trace? It looks like a small enough change, maybe a few hours work, but I'd like your feedback before I try coding it.

**Comment by GilesBathgate** [comment]
@doug-moen Bear in mind that the line number needs to come from the lexer/parser and be added to the AST. As a minimum the Function/Module name trace, should be easier to provide, by walking the context stack as you mention. (I think the need for separate ```EvalContext``` vs ```Context``` is for lexical context vs execution context, IIRC)

**Comment by doug-moen** [comment]
@GilesBathgate "Bear in mind that the line number needs to come from the lexer/parser and be added to the AST."

Yes, but IMO we don't need to add it to every single AST node. I am proposing to add it to ExpressionFunctionCall and ModuleInstantiation, which is a fairly simple change. A couple lines changed in those 2 classes, plus a few lines changed in parser.y.

**Comment by GilesBathgate** [comment]
@doug-moen Yeah, sticking it in a node be it every node, or just ExpressionFunctionCall or ModuleInstantiation is the easy bit. Its the changes in lexer.l/parser.y that I think @t-paul  was eluding to as a bigger problem.

**Comment by doug-moen** [comment]
@GilesBathgate The file name and line number are already available in parser.y as parser_source_path and lexerget_lineno(). The changes to parser.y are trivial, as I just have to pass these values as arguments to ExpressionFunctionCall::create() and ModuleInstantiation::create().

Oh, and ModuleInstantiation already contains a data member for the file name. It's used when a module has a file name argument, and needs to interpret the file name relative to the pathname of the source file containing the module call.

**Comment by GilesBathgate** [comment]
@doug-moen Hmm, yeah does seem quite simple then, I wonder what the "bigger problem" is. You may as well submit a patch @doug-moen.

**Comment by t-paul** [comment]
No, the line number ins `parser.y` are not correct, that's a long standing issue with `include` #214. Fixing that would be a huge improvement for all error reporting. Maybe it's not a big code change, but changing file handling always needs testing on all platforms, so it's a bit time consuming.

In general, the module names should be available already as this can be even queried by a user function. About the function trace, I'm not so sure yet. I'm starting to understand the relation of the various `Context` and `EvalContext` objects, adding an `EvalContext` to a `Context` feels wrong, I think EvalContext objects are supposed to be short lived at the place where the evaluation happens. @kintel can probably provide greater insight about the evaluation...

We also need to watch out for the stack frame size, otherwise we will reduce the possible recursion limit which is not great to start with.

Well, in short, all I'm saying is: There's a number of related issues that are very high on the wish list, lets do them in small steps :-).

(patches of cause welcome, but the discussion is already a very helpful)

**Comment by doug-moen** [comment]
@t-paul said "No, the line number ins parser.y are not correct, that's a long standing issue with include #214. Fixing that would be a huge improvement for all error reporting."

First, this bug is orthogonal to the assertion failure stack trace. The stack trace can be implemented independent of the bug fix. The interface for fetching the current line number would not change.

Second, this bug is easy to fix. The implementation of `include` uses a path_stack global variable to push and pop the current source file name before and after processing the file. The fix is to also push and pop the line number. It's not that scary of a change.

As for the contexts, in a function call, the EvalContext evalctx is constructed as a local variable, then Function::evaluate() is called with evalctx as an argument, and it constructs Context c as a local variable. The evalctx has a longer lifetime than c, so it is safe for c to point to evalctx.

**Comment by kintel** [comment]
I've taken note of the two suggested changes that sound sensible, and I'll review those when I review the rest of #1587.

Some thoughts (mostly @doug-moen):

In terms of improving mapping from the AST to the source code, I started on this in the [ast-refactor branch](https://github.com/openscad/openscad/tree/ast-refactor). I think it's preferable to stick a source code reference to every AST node as that may help with both error reporting and source code navigation later on. I'm thinking about e.g. dropping small icons in the editor margin to tag warnings and errors, with a mouse-over message.

I like your quick-fix to push/pop line numbers for includes. FYI: I'm considering rather making the include an AST node instead, as there shouldn't be any context from the included file which would influence later parsing. This would require some changes to the evaluator, so it's slightly longer term -> a push/pop quickfix is probably a good start.

Context handling is due for some refactoring to clean up the somewhat convoluted logic for how to handle different contexts. That might wait until we progress a few steps in the direction of OpenSCAD2 though, so we probably shouldn't wait for that before looking into stack traces.

Stack traces are generally useful for error reporting, also outside of assert(), and it might be beneficial to submit that as a separate pull request, enhancing both assert() as well as the appropriate other errors, rather than adding a blocker for merging #1587.

**Comment by dcousens** [comment]
+1 for assert(condition, msg="message")

**Comment by t-paul** [comment]
Implemented via #1827.

**Comment by ringerc** [comment]
It would make a lot of sense to utilize the stack this tracks to emit stacks for warnings too.

I'd also like to suggest an idea we use in the PostgreSQL source code, called an errcontext callback. It's a function you register in a scope that is called when unwinding the stack. It can annotate the error message with things like local parameter values. In OpenSCAD this would likely take the form of a scoped list of variables to dump the values of when unwinding the stack, like: `openscad_errcontext_params['varname1', 'varname2', ...]`.
